pkg/scheduler/framework/v1alpha1/framework.go

```diff
@@ -170,6 +189,68 @@ func (f *framework) QueueSortFunc() LessFunc {
 	return f.queueSortPlugins[0].Less
 }
 
+// RunFilterPlugins runs the set of configured filter plugins for pod on each candidate host.
+// If any of these plugins returns error on a node, the node is not suitable for running pod.
+// Meanwhile, the failure message and status are set for each non-candidate host.
+func (f *framework) RunFilterPlugins(pc *PluginContext, pod *v1.Pod,
+	nodes []*v1.Node) ([]*v1.Node, FilterFailedMap) {
+
+	// FilterPlugins is not configured
+	if len(f.filterPlugins) == 0 {
+		return nodes, nil
+	}
+
+	var (
+		mutex   sync.Mutex
+		candNum int32
+		candNode []*v1.Node
+	)
+
+	totalNum := len(nodes)
+	failedMap := make(FilterFailedMap)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	checkNode := func(i int) {
+		if i < 0 || i >= totalNum {
+			return
+		}
+		// Find NodeInfo from snapshot for running Filter in each plugin
+		nodeName := nodes[i].Name
+		nodeInfo, exist := f.nodeInfoSnapshot.NodeInfoMap[nodeName]
+		if !exist {
+			mutex.Lock()
+			failedMap[nodeName] = NewStatus(1, "Fail to find NodeInfo in snapshot")
+			mutex.Unlock()
+			return
+		}
+
+		// Check if the current pod runs on node or not
+		// If any plugin returns error, the node is filtered out
+		for _, p := range f.filterPlugins {
+			status := p.Filter(pc, pod, nodeInfo)
+			if status == nil || !status.IsSuccess() {
+				mutex.Lock()
+				failedMap[nodeName] = status
+				mutex.Unlock()
+				return
+			}
+		}
+
+		length := atomic.AddInt32(&candNum, 1)
+		if int(length) > totalNum {
```

Comment:
why do we need this? doesn't ParallelizeUntil make sure only totalNum is processed?